### DISCUSSION TITLE
Is any other numeric type other than Double not supported?

### DISCUSSION THREAD

**Discussion by meftunca** [stephenberry/glaze#1069] [Q&A] [answered]
I moved from tao json to glaze, but the move is more painful than I thought. This line confirms that no numeric type other than double is defined. Please tell me if I'm wrong, but isn't it wrong that no other value types other than double are kept?
https://github.com/stephenberry/glaze/blob/56755c224a0f8b07d95ae2c4c6eedbf531b6b28f/include/glaze/json/json_t.hpp#L22

**Comment by stephenberry** [comment, marked as answer]
This is similar to how many scripting languages work, where the default numeric type is a double. If it could be an integer or a double then reading in `1.0` vs `1` would change the type, which would make code more complex as the developer would have to constantly call `std::visit` on numeric types because changes to the input JSON could change the type.

However, in most cases using `glz::json_t` is the wrong approach to using Glaze, as it isn't strongly typed. It is usually easier (and better performing) to use C++ structs, which support explicit numeric types (like `uint16_t`, `int32_t`, `float`, `double`, etc.).

**Comment by meftunca** [comment]
@stephenberry Is there a problem with integrating my own json interface?

**Reply by stephenberry** [reply]
There should be no problem integrating your own JSON interface. You can write your own [custom serialization/deserialization](https://github.com/stephenberry/glaze/blob/main/docs/custom-serialization.md) as well for your own json struct. You can also call `.as<int>()` if with `glz::json_t` to get an integer even though it is internally stored as a double.

**Reply by meftunca** [reply]
The source of my need for json_t is to be able to handle rest api requests correctly. The main role of the Glaze library in my project is to use the beve feature. At the moment, the insufficient “json_t” feature is making my work longer :) The project I'm working on is a database project built on the collection-document model. I'm looking for the most optimized way for my project as it will be very transaction intensive.

**Reply by meftunca** [reply]
@stephenberry I revised “json_t” and unfortunately I started getting syntax_error. Can you share with me where I'm going wrong?
https://gcc.godbolt.org/z/hh78sGYrP

**Comment by stephenberry** [comment]
The reason why your code isn't working is that by adding `std::uint64_t` to the variant in the `json_t` your number type is no longer auto-deducible. The [variant-handling](https://github.com/stephenberry/glaze/blob/main/docs/variant-handling.md) documentation explains this a bit more. This is actually related to the reason why javascript doesn't directly support big integers.

Thanks so much for sharing your code for your `json_t` with more features. I'll add these to `glz::json_t` to make your life easier.

The tricky part is handling `uint64_t` (and `int64_t`). I can understand the need for the support. I think I need to make a numeric variant type that has auto-deduction for numeric types: `std::variant<double, int64_t, uint64_t>`. The best numeric type can be chosen to fit the value and allow checked conversions to whatever type the developer wants. I've opened an issue for this here: #1079

**Reply by meftunca** [reply]
Dude, you're awesome. My project really needed this :) In the future I will share with you a data management project based on beve for your appreciation. It will have indexing and filtering features like orc,parquet and avro. I believe we can show everyone the power of beve and glaze library :D

**Reply by stephenberry** [reply]
Sweet, I'll try to get these features merged soon. Your project sounds awesome and I'd love to see it in the future!
